cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/TransformExecutorFactory.java

```diff
@@ -77,23 +85,88 @@ protected abstract TrackedTransform getTransformation(String pluginType, String
+        outputSchemas.put(stageName, stageInfo.getOutputSchema());
         perStageInputSchemas.put(stageName, stageInfo.getInputSchemas());
-        // Wrap each transformation so that each stage is emitting stageName along with the record
-        transformations.put(stageName,
-                            new TransformDetail(getTransformation(pluginType, stageName),
-                                                pipeline.getStageOutputs(stageName)));
       }
     }
 
+    // recursively set BatchTransformDetail for all the stages
+    for (String source : sources) {
+      setBatchTransformDetail(pipeline, source, transformations, transformErrorSinkMap, outputWriter);
+    }
+
     // sourceStageName will be null in reducers, so need to handle that case
     Set<String> startingPoints = (sourceStageName == null) ? pipeline.getSources() : Sets.newHashSet(sourceStageName);
-    return new TransformExecutor<>(transformations, startingPoints);
+    return new BatchTransformExecutor<>(transformations, startingPoints);
+  }
+
+  private <KEY_OUT, VAL_OUT> void setBatchTransformDetail(PipelinePhase pipeline, String stageName,
+                                                          Map<String, BatchTransformDetail> transformations,
+                                                          Map<String, ErrorOutputWriter<Object, Object>>
+                                                            transformErrorSinkMap,
+                                                          OutputWriter<KEY_OUT, VAL_OUT> outputWriter)
+    throws Exception {
+    if (pipeline.getSinks().contains(stageName)) {
+      String pluginType = pipeline.getStage(stageName).getPluginType();
+      // If there is a connector sink/ joiner at the end of pipeline, do not remove stage name. This is needed to save
+      // stageName along with the record in connector sink and joiner takes input along with stageName
+      boolean removeStageName = !(pluginType.equalsIgnoreCase(Constants.CONNECTOR_TYPE) ||
+        pluginType.equalsIgnoreCase(BatchJoiner.PLUGIN_TYPE));
+      transformations.put(stageName, new BatchTransformDetail(stageName, removeStageName,
+                                                              getTransformation(pluginType, stageName),
+                                                              new SinkEmitter<>(stageName, outputWriter)));
+      return;
+    }
+
+    for (String output : pipeline.getDag().getNodeOutputs(stageName)) {
+      setBatchTransformDetail(pipeline, output, transformations, transformErrorSinkMap, outputWriter);
+
+      // Add new transformation for output stage if the transfomation for stageName already exists
+      if (transformations.containsKey(stageName)) {
+        BatchEmitter<BatchTransformDetail> emitter = transformations.get(stageName).getEmitter();
+        Map<String, BatchTransformDetail> nextStages = emitter.getNextStages();
+        if (nextStages != null && !nextStages.containsKey(output)) {
```

Comment:
ok I see its in the loop because of this. I don't think you actually need to check the next stages right? Can't you just add the TransformDetail directly?